R/nullModelApproxSE.R

```diff
@@ -0,0 +1,220 @@
+                         male.diploid = male.diploid, genome.build = genome.build, verbose = verbose)
+     return(tab)
+}
+	
+
+updateNullModApproxSE <- function(null.model, tab){
+     # rbind a list of tables
+     if(class(tab) == 'list') tab <- data.table::rbindlist(tab)
+
+     # compute the mean of se.ratio
+     r <- mean(tab$se.ratio)
+
+     # compute SE(r) and check if enough variants were included
+     r.se <- sd(tab$se.ratio)/sqrt(nrow(tab))
+     val <- r.se/r
+     message(paste('mean se.ratio: r = ', r))
+     message(paste('SE(r)/r = ', val))
+     if(val > 0.0025) warning(paste('It is recommended that SE(r)/r be < 0.0025. It is suggested to increase the number of variants in tab; try at least', 
+          round(nrow(tab)*(val/0.0025)^2, 0), 'variants'))
+
+     # update the null model
+     null.model <- nullModelSmall(null.model)
+
+     null.model$se.correction <- r
+     null.model$se.table <- tab
+     return(null.model)
+}
+
+
+## function to get a random sample of variants with a minimum MAC
+## this is called by updateNullModelApproxScoreSE
+setGeneric(".selectRandomVars", function(gdsobj, ...) standardGeneric(".selectRandomVars"))
+
+setMethod(".selectRandomVars",
+          "SeqVarData",
+          function(gdsobj, 
+                   sample.id = NULL, 
+                   nvar = 100, 
+                   min.mac = 20, 
+                   verbose = TRUE){
+
+               # number of variants in the GDS object
+               nvar.gdsobj <- SeqArray::seqSummary(gdsobj, verbose = FALSE)$num.variant
+               if(nvar.gdsobj < nvar) stop('requested more variants than available in gdsobj')
+
+               # filter to sample set in null.model
+               SeqArray::seqSetFilter(gdsobj, sample.id = sample.id, verbose = verbose)
+
+               out <- NULL
+               while(length(out) < nvar){
+                    # sample variants
+                    var.rand <- sort(sample.int(nvar.gdsobj, size = nvar, replace = FALSE))
+                    # filter to random sample
+                    SeqArray::seqSetFilter(gdsobj, variant.sel = var.rand, verbose = verbose)
+                    # filter to MAC threshold
+                    if(min.mac > 0) SeqArray::seqSetFilterCond(gdsobj, mac = min.mac, verbose = verbose)
+                    # collect selected variants
+                    out <- sort(unique(c(out, seqGetData(gdsobj, 'variant.id'))))
+               }
+
+               # sample down to number requested
+               out <- sample(out, size = nvar, replace = FALSE)
+               return(out)
+               })
+
+
+## function to calculate both the true and fast Score.SE for a specified set of variants
+## computes the ratio of the two Score.SE estimates for calculating the correction factor
+## this is called by updateNullModelApproxScoreSE
+setGeneric(".calcScore", function(gdsobj, ...) standardGeneric(".calcScore"))
+
```

Comment:
made it so that this function also returns the score value. might be useful if a user just wants to quickly compute the score for a few variants from a GDS file and null model.